cinder/volume/manager.py

```diff
@@ -586,6 +586,50 @@ def _init_host(self, added_to_cluster=None, **kwargs):
         super(VolumeManager, self).init_host(added_to_cluster=added_to_cluster,
                                              **kwargs)
 
+    @coordination.synchronized('volume-stats')
+    def recount_host_volumes(self, context):
+        """Recount the number of volumes and allocated capacity."""
+        ctxt = context.elevated()
+        LOG.info("Recounting Allocated capacity")
+
+        # Zero stats
+        self.stats['pools'] = {}
+        self.stats.update({'allocated_capacity_gb': 0})
+
+        # Batch retrieval volumes and snapshots
+        num_vols, req_range = None, [0]
+        req_limit = CONF.init_host_max_objects_retrieval
+        use_batch_objects_retrieval = req_limit > 0
+
+        if use_batch_objects_retrieval:
+            # Get total number of volumes
+            num_vols, __, __ = self._get_my_volumes_summary(ctxt)
+            # Make batch request loop counter
+            req_range = range(0, num_vols, req_limit)
+
+        for req_offset in req_range:
+            # Retrieve 'req_limit' number of objects starting from
+            # 'req_offset' position
+            volumes = None
+            if use_batch_objects_retrieval:
+                if req_offset < num_vols:
+                    volumes = self._get_my_volumes(ctxt,
+                                                   limit=req_limit,
+                                                   offset=req_offset)
+                else:
+                    volumes = objects.VolumeList()
+            # or retrieve all volumes and snapshots per single request
+            else:
+                volumes = self._get_my_volumes(ctxt)
+
+            for volume in volumes:
+                # Account for volumes that have been provisioned already
+                if volume['host']:
+                    # calculate allocated capacity for driver
+                    self._count_allocated_capacity(ctxt, volume)
+
+            del volumes
```

Comment:
This is a memory usage optimization that originated in the original code.  Some times there is a max object fetch limit specified in cinder.conf to prevent high memory usage and this for loop can iterate over N volumes where N is the max object fetch limit.  deleting these explicitly will free up memory between loop interations.   This was in the original code that I pulled from init_host.

Reply:
This is to explicitly trigger garbage-collection, right? Because the `volumes` should be out of scope directly in the next loop as I would understand it.

Anyways, thank you for the explanation.